src/hevm/src/EVM/SolidityABI.hs

```diff
@@ -0,0 +1,153 @@
+  name  <- (skipSpaces *> identifier)
+  types <- (skipSpaces *> parseTypes)
+  -- n.b. modifier arguments are not supported
+  modifiers <- option [] $
+    space *> sepBy1 (identifier `excluding` ["returns"]) (many1 (char ' '))
+  -- allow us to parse f(inputs)(outputs) with or without the "returns"
+  returns <- option [] $
+    optional (space *> string "returns") *> skipSpaces *> parseTypes
+  skipSpaces
+  optional (char ';')
+  skipSpaces
+  eof
+  let mutability =
+        case (flip elem modifiers) <$> ["view", "pure", "payable"] of
+          [False, False, False] -> NonPayable
+          [True,  False, False] -> View
+          [False, True,  False] -> Pure
+          [False, False, True ] -> Payable
+          _ -> error "overspecified mutability"
+  return $ Method
+    (fst <$> returns)
+    (fst <$> types)
+    (pack name)
+    (pack $ name <> (show (AbiTupleType (V.fromList (snd <$> fst <$> types)))))
+    mutability
+
+parseError :: ReadP SolError
+parseError = do
+  (skipSpaces *> string "error" *> space)
+  name  <- (skipSpaces *> identifier)
+  types <- (skipSpaces *> parseTypes)
+  skipSpaces
+  optional (char ';')
+  skipSpaces
+  eof
+  return $ SolError (pack name) (snd <$> fst <$> types)
+
+data Location
+  = MemoryLocation
+  | CalldataLocation
+  | IndexedLocation
+  | NoLocation
+
+parseEvent :: ReadP Event
+parseEvent = do
+  skipSpaces *> string "event" *> space
+  name  <- (skipSpaces *> identifier)
+  types <- (skipSpaces *> parseTypes)
+  anon  <- option "" (space *> string "anonymous")
+  skipSpaces
+  optional (char ';')
+  skipSpaces
+  eof
+  let
+    anonymous = \case
+      "anonymous" -> Anonymous
+      "" -> NotAnonymous
+    indexed = \case
+      ((argname, typ), IndexedLocation) -> (argname, typ, Indexed)
+      ((argname, typ), _)               -> (argname, typ, NotIndexed)
+  return $ EVM.ABI.Event (pack name) (anonymous anon) (indexed <$> types)
+
+parseTypes :: ReadP [((Text, AbiType), Location)]
+parseTypes = between
+  (char '(' *> skipSpaces)
+  (skipSpaces <* char ')')
+  (sepBy parseType (skipSpaces <* char ',' *> skipSpaces))
+
+parseType :: ReadP ((Text, AbiType), Location)
+parseType = do
+  t <- parseBasicType
+  s <- many (between (char '[') (char ']') (munch isDigit))
+  loc  <- option "" (space *> location)
+  name <- option "" (space *> argName)
+  return ((pack name, foldl makeArray t s), locate loc)
+  where
+    location = (string "memory" +++ string "calldata" +++ string "indexed")
+    locate = \case
+      "memory"   -> MemoryLocation
+      "calldata" -> CalldataLocation
+      "indexed"  -> IndexedLocation
+      ""         -> NoLocation
+      _          -> error "unknown location"
+
+    argName = identifier `excluding` ["memory", "calldata", "indexed"]
+
+    makeArray :: AbiType -> String -> AbiType
+    makeArray t "" = AbiArrayDynamicType t
+    makeArray t s  = AbiArrayType (read s) t
+
+    parseBasicType :: ReadP AbiType
```

Comment:
this would be nice to have in ABI.hs next to the show instance for AbiType. We can also easily quickcheck the roundtrip